src/Coordination/KeeperSnapshotManagerS3.cpp

```diff
@@ -0,0 +1,293 @@
+            std::move(headers),
+            auth_settings.use_environment_credentials.value_or(false),
+            auth_settings.use_insecure_imds_request.value_or(false));
+
+        auto new_client = std::make_shared<KeeperSnapshotManagerS3::S3Configuration>(std::move(new_uri), std::move(auth_settings), std::move(client));
+
+        client_lock.lock();
+        snapshot_s3_client = std::move(new_client);
+        client_lock.unlock();
+        LOG_INFO(log, "S3 client was updated");
+    }
+    catch (...)
+    {
+        LOG_ERROR(log, "Failed to create an S3 client for snapshots");
+        tryLogCurrentException(__PRETTY_FUNCTION__);
+    }
+}
+std::shared_ptr<KeeperSnapshotManagerS3::S3Configuration> KeeperSnapshotManagerS3::getSnapshotS3Client() const
+{
+    std::lock_guard lock{snapshot_s3_client_mutex};
+    return snapshot_s3_client;
+}
+
+void KeeperSnapshotManagerS3::snapshotS3Thread()
+{
+    setThreadName("KeeperS3SnpT");
+
+    auto uuid = UUIDHelpers::generateV4();
+    while (!shutdown_called)
+    {
+        std::string snapshot_path;
+        if (!snapshots_s3_queue.pop(snapshot_path))
+            break;
+
+        if (shutdown_called)
+            break;
+
+        try
+        {
+            auto s3_client = getSnapshotS3Client();
+            if (s3_client == nullptr)
+                continue;
+
+            LOG_INFO(log, "Will try to upload snapshot on {} to S3", snapshot_path);
+            ReadBufferFromFile snapshot_file(snapshot_path);
+
+            S3Settings::ReadWriteSettings read_write_settings;
+            read_write_settings.upload_part_size_multiply_parts_count_threshold = 10000;
+
+            const auto create_writer = [&](const auto & key)
+            {
+                return WriteBufferFromS3
+                {
+                    s3_client->client,
+                    s3_client->uri.bucket,
+                    key,
+                    read_write_settings
+                };
+            };
+
+            auto snapshot_name = fs::path(snapshot_path).filename().string();
+            auto lock_file = fmt::format(".{}_LOCK", snapshot_name);
+
+            const auto file_exists = [&](const auto & key)
+            {
+                Aws::S3::Model::HeadObjectRequest request;
+                request.SetBucket(s3_client->uri.bucket);
+                request.SetKey(key);
+                auto outcome = s3_client->client->HeadObject(request);
+
+                if (outcome.IsSuccess())
+                    return true;
+
+                const auto & error = outcome.GetError();
+                if (error.GetErrorType() != Aws::S3::S3Errors::NO_SUCH_KEY && error.GetErrorType() != Aws::S3::S3Errors::RESOURCE_NOT_FOUND)
+                    throw S3Exception(error.GetErrorType(), "Failed to verify existence of lock file: {}", error.GetMessage());
+
+                return false;
+            };
+
+            if (file_exists(snapshot_name))
```

Comment:
how do we sure that for every snapshot we generate different name? Otherwise only first snapshot will be saved
Also, if we generate unique name for each snapshot, who is taking care for removing aged snapshots?

Reply:
Please add comment about 2 leaders and concurrent writes